driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -387,21 +436,71 @@ private void maybeSpawnNewConnection() {
                 break;
         }
 
-        logger.debug("Creating new connection on busy pool to {}", host);
         manager.blockingExecutor().submit(newConnectionTask);
     }
 
     void replaceDefunctConnection(final PooledConnection connection) {
-        if (connection.markForTrash.compareAndSet(false, true))
+        if (connection.state.compareAndSet(OPEN, GONE))
             open.decrementAndGet();
-        connections.remove(connection);
+        if (connections.remove(connection))
+            manager.blockingExecutor().submit(new Runnable() {
+                @Override
+                public void run() {
+                    addConnectionIfUnderMaximum();
+                }
+            });
         connection.closeAsync();
-        manager.blockingExecutor().submit(new Runnable() {
-            @Override
-            public void run() {
-                addConnectionIfUnderMaximum();
+    }
+
+    void cleanupIdleConnections(long now) {
+        if (isClosed())
+            return;
+
+        shrinkIfBelowCapacity();
+        cleanupTrash(now);
+    }
+
+    /** If we have more active connections than needed, trash some of them */
+    private void shrinkIfBelowCapacity() {
+        int currentLoad = maxTotalInFlight.getAndSet(0);
+
+        int needed = currentLoad / StreamIdGenerator.MAX_STREAM_PER_CONNECTION + 1;
+        if (currentLoad % StreamIdGenerator.MAX_STREAM_PER_CONNECTION > options().getMaxSimultaneousRequestsPerConnectionThreshold(hostDistance))
+            needed += 1;
+        needed = Math.max(needed, options().getCoreConnectionsPerHost(hostDistance));
+        int actual = open.get();
+        int toTrash = Math.max(0, actual - needed);
+
+        logger.trace("Current inFlight = {}, {} connections needed, {} connections available, trashing {}",
+            currentLoad, needed, actual, toTrash);
+
+        if (toTrash <= 0)
+            return;
+
+        for (PooledConnection connection : connections)
+            if (trashConnection(connection)) {
```

Comment:
I've given this some thought, and I think we *do* want to count trashed connections towards `totalInFlight`. Our goal is to estimate the current load on the pool. Whether some of these in-flight requests are executing on a trashed connection is irrelevant.

For instance, assume the app produces a constant load of 250 concurrent requests. At some point we have 3 connections and the load is distributed as c1:100, c2: 100, c3: 50. Now for some reason c3 gets trashed. If we ignore the load on trashed connections, we'll now estimate `totalInFlight` to 200, and consider that the current size is enough. But the app still has 250 concurrent requests, so that's what we should plan for.